File: src/spread_analyzer.py
"""
价差分析模块
负责计算价差、收敛区间和生成交易信号
"""
import numpy as np
import logging
from typing import Dict, List, Optional, Tuple
from datetime import datetime, timedelta
from collections import defaultdict
import config

logger = logging.getLogger(__name__)


class SpreadAnalyzer:
    """价差分析器"""
    
    def __init__(self):
        # 存储历史价差数据: {(dex1, dex2, symbol): [spread1, spread2, ...]}
        self.spread_history: Dict[Tuple[str, str, str], List[float]] = defaultdict(list)
        # 存储历史价格数据: {(dex, symbol): [(timestamp, price), ...]}
        self.price_history: Dict[Tuple[str, str], List[Tuple[datetime, float]]] = defaultdict(list)
    
    def calculate_spread(self, price1: float, price2: float) -> float:
        """
        计算价差（绝对值）
        
        Args:
            price1: 价格1
            price2: 价格2
            
        Returns:
            价差
        """
        return abs(price1 - price2)
    
    def calculate_spread_percentage(self, price1: float, price2: float) -> float:
        """
        计算价差百分比
        
        Args:
            price1: 价格1
            price2: 价格2
            
        Returns:
            价差百分比
        """
        min_price = min(price1, price2)
        if min_price == 0:
            return 0
        return abs(price1 - price2) / min_price
    
    def update_spread_history(self, dex1: str, dex2: str, symbol: str, spread: float):
        """
        更新价差历史数据
        
        Args:
            dex1: DEX1名称
            dex2: DEX2名称
            symbol: 币种符号
            spread: 价差
        """
        key = (dex1, dex2, symbol)
        self.spread_history[key].append(spread)
        
        # 只保留最近24小时的数据（假设每秒更新一次，约86400条）
        max_history = config.SPREAD_CONFIG['convergence_window_hours'] * 3600
        if len(self.spread_history[key]) > max_history:
            self.spread_history[key] = self.spread_history[key][-max_history:]
    
    def calculate_convergence_range(self, dex1: str, dex2: str, symbol: str) -> Optional[Dict]:
        """
        计算价差收敛区间
        
        Args:
            dex1: DEX1名称
            dex2: DEX2名称
            symbol: 币种符号
            
        Returns:
            收敛区间信息: {mean, std, lower, upper}
        """
        key = (dex1, dex2, symbol)
        history = self.spread_history.get(key, [])
        
        if len(history) < 10:  # 至少需要10个数据点
            return None
        
        mean = np.mean(history)
        std = np.std(history)
        
        # 使用2倍标准差作为收敛区间
        return {
            'mean': mean,
            'std': std,
            'lower': mean - 2 * std,
            'upper': mean + 2 * std,
            'count': len(history),
        }
    
    def analyze_spread_opportunity(
        self, 
        dex1: str, 
        dex2: str, 
        symbol: str, 
        price1: float, 
        price2: float,
        binance_price: Optional[float] = None
    ) -> Dict:
        """
        分析价差套利机会
        
        Args:
            dex1: DEX1名称
            dex2: DEX2名称
            symbol: 币种符号
            price1: DEX1价格
            price2: DEX2价格
            binance_price: 币安价格（可选）
            
        Returns:
            分析结果
        """
        # 计算价差
        spread = self.calculate_spread(price1, price2)
        spread_pct = self.calculate_spread_percentage(price1, price2)
        
        # 更新历史数据
        self.update_spread_history(dex1, dex2, symbol, spread)
        
        # 计算收敛区间
        convergence = self.calculate_convergence_range(dex1, dex2, symbol)
        
        # 判断价差方向
        if price1 > price2:
            high_dex = dex1
            low_dex = dex2
            high_price = price1
            low_price = price2
        else:
            high_dex = dex2
            low_dex = dex1
            high_price = price2
            low_price = price1
        
        result = {
            'dex1': dex1,
            'dex2': dex2,
            'symbol': symbol,
            'price1': price1,
            'price2': price2,
            'spread': spread,
            'spread_percentage': spread_pct,
            'high_dex': high_dex,
            'low_dex': low_dex,
            'high_price': high_price,
            'low_price': low_price,
            'convergence': convergence,
        }
        
        return result
    
    def calculate_trading_cost(
        self, 
        dex1: str, 
        dex2: str, 
        amount: float, 
        use_maker: bool = True
    ) -> float:
        """
        计算交易成本
        
        Args:
            dex1: DEX1名称
            dex2: DEX2名称
            amount: 交易金额
            use_maker: 是否使用maker订单
            
        Returns:
            总手续费（开仓+平仓）
        """
        fee_type = 'maker' if use_maker else 'taker'
        
        fee1 = config.DEX_CONFIG[dex1]['fees'][fee_type] * amount
        fee2 = config.DEX_CONFIG[dex2]['fees'][fee_type] * amount
        
        # 开仓和平仓都需要手续费
        total_cost = (fee1 + fee2) * 2
        
        return total_cost
    
    def check_arbitrage_opportunity(
        self,
        dex1: str,
        dex2: str,
        symbol: str,
        price1: float,
        price2: float,
        amount: float = 1000,
        use_maker: bool = True
    ) -> Dict:
        """
        检查套利机会
        
        Args:
            dex1: DEX1名称
            dex2: DEX2名称
            symbol: 币种符号
            price1: DEX1价格
            price2: DEX2价格
            amount: 交易金额
            use_maker: 是否使用maker订单
            
        Returns:
            套利机会分析结果
        """
        # 分析价差
        spread_analysis = self.analyze_spread_opportunity(dex1, dex2, symbol, price1, price2)
        
        # 计算交易成本（基于交易金额）
        total_cost = self.calculate_trading_cost(dex1, dex2, amount, use_maker)
        
        # 计算预期价差利润（基于交易金额）
        # 价差利润 = (当前价差 - 预期收敛价差) / 价格 × 交易金额
        convergence = spread_analysis['convergence']
        spread_pct = spread_analysis['spread_percentage']
        min_price = min(price1, price2)
        
        if convergence:
            # 预期价差收敛到均值，利润 = (当前价差百分比 - 均值价差百分比) × 交易金额
            mean_spread_pct = convergence['mean'] / min_price if min_price > 0 else 0
            expected_spread_pct = spread_pct - mean_spread_pct
            expected_profit = expected_spread_pct * amount
        else:
            # 没有历史数据，保守估计：假设价差收敛50%
            expected_profit = spread_pct * amount * 0.5
        
        # 判断是否有利可图（预期利润需要覆盖成本）
        net_profit = expected_profit - total_cost
        is_profitable = net_profit > 0
        
        # 判断是否超出阈值
        min_threshold = config.SPREAD_CONFIG['min_spread_threshold']
        max_threshold = config.SPREAD_CONFIG['max_spread_threshold']
        
        within_threshold = min_threshold <= spread_pct <= max_threshold
        
        # 判断是否超出收敛区间（如果没有历史数据，使用价差百分比阈值）
        beyond_convergence = False
        if convergence:
            # 有历史数据：判断是否超出收敛区间上界
            beyond_convergence = spread_analysis['spread'] > convergence['upper']
        else:
            # 没有历史数据：使用价差百分比阈值判断
            # 如果价差超过最小阈值，认为可能存在机会
            beyond_convergence = spread_pct >= min_threshold
        
        result = {
            **spread_analysis,
            'amount': amount,
            'total_cost': total_cost,
            'expected_profit': expected_profit,
            'net_profit': net_profit,
            'is_profitable': is_profitable,
            'within_threshold': within_threshold,
            'beyond_convergence': beyond_convergence,
            'signal': self._generate_signal(
                spread_analysis, 
                is_profitable, 
                within_threshold, 
                beyond_convergence,
                convergence is not None
            ),
            'debug_info': {
                'has_convergence': convergence is not None,
                'convergence_mean': convergence['mean'] if convergence else None,
                'convergence_upper': convergence['upper'] if convergence else None,
                'current_spread': spread_analysis['spread'],
                'spread_pct': spread_pct,
            }
        }
        
        return result
    
    def _generate_signal(
        self,
        spread_analysis: Dict,
        is_profitable: bool,
        within_threshold: bool,
        beyond_convergence: bool,
        has_convergence_data: bool = False
    ) -> Optional[Dict]:
        """
        生成交易信号
        
        Args:
            spread_analysis: 价差分析结果
            is_profitable: 是否有利可图
            within_threshold: 是否在阈值范围内
            beyond_convergence: 是否超出收敛区间
            has_convergence_data: 是否有收敛区间数据
            
        Returns:
            交易信号或None
        """
        # 放宽信号生成条件：
        # 1. 如果有收敛区间数据：需要超出收敛区间且有利可图
        # 2. 如果没有收敛区间数据：只要价差在阈值范围内且有利可图即可
        
        if has_convergence_data:
            # 有历史数据：需要超出收敛区间
            condition = beyond_convergence and is_profitable and within_threshold
        else:
            # 没有历史数据：只要价差在阈值范围内且有利可图
            condition = is_profitable and within_threshold
        
        if condition:
            return {
                'action': 'OPEN',
                'high_dex': spread_analysis['high_dex'],
                'high_side': 'SHORT',
                'high_price': spread_analysis['high_price'],
                'low_dex': spread_analysis['low_dex'],
                'low_side': 'LONG',
                'low_price': spread_analysis['low_price'],
                'spread': spread_analysis['spread'],
                'spread_percentage': spread_analysis['spread_percentage'],
            }
        
        # 平仓信号：价差收敛到正常区间（如果有持仓的话）
        # 这里需要结合持仓状态判断，暂时返回None
        # 实际使用时需要传入持仓信息
        
        return None

File: chrome-extension/lib/spread-analyzer.js
class SpreadAnalyzer {
  constructor() {
    this.spreadHistory = {};
    this.priceHistory = {};
  }


  async loadHistory() {
    try {
      const result = await chrome.storage.local.get(['spreadHistory', 'priceHistory']);
      if (result.spreadHistory) {
        this.spreadHistory = result.spreadHistory;
      }
      if (result.priceHistory) {
        this.priceHistory = result.priceHistory;
      }
      console.log('历史数据已加载');
    } catch (error) {
      console.error('加载历史数据失败:', error);
    }
  }

 
  async saveHistory() {
    try {
      await chrome.storage.local.set({
        spreadHistory: this.spreadHistory,
        priceHistory: this.priceHistory,
      });
    } catch (error) {
      console.error('保存历史数据失败:', error);
    }
  }

  /**
   * 计算价差（绝对值）
   * @param {number} price1
   * @param {number} price2
   * @returns {number}
   */
  calculateSpread(price1, price2) {
    return Math.abs(price1 - price2);
  }

  /**
   * 计算价差百分比
   * @param {number} price1
   * @param {number} price2
   * @returns {number}
   */
  calculateSpreadPercentage(price1, price2) {
    const minPrice = Math.min(price1, price2);
    if (minPrice === 0) return 0;
    return Math.abs(price1 - price2) / minPrice;
  }

  /**
   * 更新价差历史数据
   * @param {string} dex1
   * @param {string} dex2
   * @param {string} symbol
   * @param {number} spread
   */
  updateSpreadHistory(dex1, dex2, symbol, spread) {
    const key = `${dex1}|${dex2}|${symbol}`;
    if (!this.spreadHistory[key]) {
      this.spreadHistory[key] = [];
    }
    this.spreadHistory[key].push(spread);

    // 只保留最近24小时的数据
    const maxHistory = CONFIG.SPREAD_CONFIG.convergence_window_hours * 3600;
    if (this.spreadHistory[key].length > maxHistory) {
      this.spreadHistory[key] = this.spreadHistory[key].slice(-maxHistory);
    }
  }

  /**
   * 计算价差收敛区间
   * @param {string} dex1
   * @param {string} dex2
   * @param {string} symbol
   * @returns {Object|null} {mean, std, lower, upper, count}
   */
  calculateConvergenceRange(dex1, dex2, symbol) {
    const key = `${dex1}|${dex2}|${symbol}`;
    const history = this.spreadHistory[key] || [];

    if (history.length < 10) {
      return null;
    }

    const mean = history.reduce((sum, val) => sum + val, 0) / history.length;
    const variance = history.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / history.length;
    const std = Math.sqrt(variance);

    return {
      mean: mean,
      std: std,
      lower: mean - 2 * std,
      upper: mean + 2 * std,
      count: history.length,
    };
  }

  /**
   * 分析价差套利机会
   * @param {string} dex1
   * @param {string} dex2
   * @param {string} symbol
   * @param {number} price1 - DEX1价格
   * @param {number} price2 - DEX2价格
   * @returns {Object} 分析结果
   */
  analyzeSpreadOpportunity(dex1, dex2, symbol, price1, price2) {
    // 计算价差
    const spread = this.calculateSpread(price1, price2);
    const spreadPct = this.calculateSpreadPercentage(price1, price2);

    // 更新历史数据
    this.updateSpreadHistory(dex1, dex2, symbol, spread);

    // 计算收敛区间
    const convergence = this.calculateConvergenceRange(dex1, dex2, symbol);

    // 判断价差方向
    let highDex, lowDex, highPrice, lowPrice;
    if (price1 > price2) {
      highDex = dex1;
      lowDex = dex2;
      highPrice = price1;
      lowPrice = price2;
    } else {
      highDex = dex2;
      lowDex = dex1;
      highPrice = price2;
      lowPrice = price1;
    }

    return {
      dex1,
      dex2,
      symbol,
      price1,
      price2,
      spread,
      spread_percentage: spreadPct,
      high_dex: highDex,
      low_dex: lowDex,
      high_price: highPrice,
      low_price: lowPrice,
      convergence,
    };
  }

  /**
   * 计算交易成本
   * @param {string} dex1
   * @param {string} dex2
   * @param {number} amount - 交易金额
   * @param {boolean} useMaker - 是否使用maker订单
   * @returns {number} 总手续费（开仓+平仓）
   */
  calculateTradingCost(dex1, dex2, amount, useMaker = true) {
    const feeType = useMaker ? 'maker' : 'taker';

    const fee1 = CONFIG.DEX_CONFIG[dex1].fees[feeType] * amount;
    const fee2 = CONFIG.DEX_CONFIG[dex2].fees[feeType] * amount;

    // 开仓和平仓都需要手续费
    return (fee1 + fee2) * 2;
  }

  /**
   * 检查套利机会
   * @param {string} dex1
   * @param {string} dex2
   * @param {string} symbol
   * @param {number} price1
   * @param {number} price2
   * @param {number} amount - 交易金额
   * @param {boolean} useMaker
   * @returns {Object} 套利机会分析结果
   */
  checkArbitrageOpportunity(dex1, dex2, symbol, price1, price2, amount = 1000, useMaker = true) {
    // 分析价差
    const spreadAnalysis = this.analyzeSpreadOpportunity(dex1, dex2, symbol, price1, price2);

    // 计算交易成本
    const totalCost = this.calculateTradingCost(dex1, dex2, amount, useMaker);

    // 计算预期价差利润
    const convergence = spreadAnalysis.convergence;
    const spreadPct = spreadAnalysis.spread_percentage;
    const minPrice = Math.min(price1, price2);

    let expectedProfit;
    if (convergence) {
      const meanSpreadPct = minPrice > 0 ? convergence.mean / minPrice : 0;
      const expectedSpreadPct = spreadPct - meanSpreadPct;
      expectedProfit = expectedSpreadPct * amount;
    } else {
      // 没有历史数据，保守估计
      expectedProfit = spreadPct * amount * 0.5;
    }

    // 判断是否有利可图
    const netProfit = expectedProfit - totalCost;
    const isProfitable = netProfit > 0;

    // 判断是否在阈值范围内
    const minThreshold = CONFIG.SPREAD_CONFIG.min_spread_threshold;
    const maxThreshold = CONFIG.SPREAD_CONFIG.max_spread_threshold;
    const withinThreshold = minThreshold <= spreadPct && spreadPct <= maxThreshold;

    // 判断是否超出收敛区间
    let beyondConvergence = false;
    if (convergence) {
      beyondConvergence = spreadAnalysis.spread > convergence.upper;
    } else {
      beyondConvergence = spreadPct >= minThreshold;
    }

    const result = {
      ...spreadAnalysis,
      amount,
      total_cost: totalCost,
      expected_profit: expectedProfit,
      net_profit: netProfit,
      is_profitable: isProfitable,
      within_threshold: withinThreshold,
      beyond_convergence: beyondConvergence,
      signal: this._generateSignal(
        spreadAnalysis,
        isProfitable,
        withinThreshold,
        beyondConvergence,
        convergence !== null
      ),
      debug_info: {
        has_convergence: convergence !== null,
        convergence_mean: convergence ? convergence.mean : null,
        convergence_upper: convergence ? convergence.upper : null,
        current_spread: spreadAnalysis.spread,
        spread_pct: spreadPct,
      },
    };

    return result;
  }

  /**
   * 生成交易信号
   * @param {Object} spreadAnalysis - 价差分析结果
   * @param {boolean} isProfitable
   * @param {boolean} withinThreshold
   * @param {boolean} beyondConvergence
   * @param {boolean} hasConvergenceData
   * @returns {Object|null} 交易信号
   */
  _generateSignal(spreadAnalysis, isProfitable, withinThreshold, beyondConvergence, hasConvergenceData = false) {
    const convergence = spreadAnalysis.convergence;
    const currentSpread = spreadAnalysis.spread;

    if (convergence) {
      const meanSpread = convergence.mean;
      const upperSpread = convergence.upper;

      // 开仓阈值
      const openThreshold = upperSpread * CONFIG.SPREAD_CONFIG.open_spread_multiplier;

      // 平仓阈值
      const closeThreshold = meanSpread * CONFIG.SPREAD_CONFIG.close_spread_multiplier;

      // 平仓信号
      if (currentSpread <= closeThreshold) {
        return {
          action: 'CLOSE',
          high_dex: spreadAnalysis.high_dex,
          high_side: 'LONG',
          low_dex: spreadAnalysis.low_dex,
          low_side: 'SHORT',
          current_spread: currentSpread,
          close_threshold: closeThreshold,
          convergence_mean: meanSpread,
        };
      }

      // 开仓信号
      if (currentSpread >= openThreshold && isProfitable && withinThreshold) {
        return {
          action: 'OPEN',
          high_dex: spreadAnalysis.high_dex,
          high_side: 'SHORT',
          high_price: spreadAnalysis.high_price,
          low_dex: spreadAnalysis.low_dex,
          low_side: 'LONG',
          low_price: spreadAnalysis.low_price,
          spread: currentSpread,
          spread_percentage: spreadAnalysis.spread_percentage,
          open_threshold: openThreshold,
        };
      }
    } else {
      // 没有收敛区间数据
      const spreadPct = spreadAnalysis.spread_percentage;
      const minThreshold = CONFIG.SPREAD_CONFIG.min_spread_threshold;
      const maxThreshold = CONFIG.SPREAD_CONFIG.max_spread_threshold;

      // 平仓信号
      if (spreadPct < minThreshold) {
        return {
          action: 'CLOSE',
          high_dex: spreadAnalysis.high_dex,
          high_side: 'LONG',
          low_dex: spreadAnalysis.low_dex,
          low_side: 'SHORT',
          current_spread: currentSpread,
          reason: '价差过小',
        };
      }

      // 开仓信号
      if (minThreshold <= spreadPct && spreadPct <= maxThreshold && isProfitable) {
        return {
          action: 'OPEN',
          high_dex: spreadAnalysis.high_dex,
          high_side: 'SHORT',
          high_price: spreadAnalysis.high_price,
          low_dex: spreadAnalysis.low_dex,
          low_side: 'LONG',
          low_price: spreadAnalysis.low_price,
          spread: currentSpread,
          spread_percentage: spreadPct,
        };
      }
    }

    return null;
  }
}
